File: Programacao_Funcional/alta_ordem/codes/isprime.cpp
#include <bits/stdc++.h>

using namespace std;

bool is_prime(int n)
{
    if (n < 2)
        return false;

    if (n == 2)
        return true;

    if (n % 2 == 0)
        return false;

    for (int d = 3; d * d <= n; d += 2)
        if (n % d == 0)
            return false;

    return true;
}

int main()
{
    int n;
    cin >> n;

    cout << n << " eh primo? " << is_prime(n) << '\n';

    return 0;
}

File: Programacao_Funcional/alta_ordem/codes/isprime.hs
is_prime 2 = True
is_prime n  | n < 2         = False
            | mod n 2 == 0  = False  
            | otherwise     = test n 3
            where test n d | d * d > n      = True
                           | mod n d == 0   = False
                           | otherwise      = test n (d + 2)

main = do
    n <- readLn :: IO Int
    putStrLn $ show n ++ " eh primo? " ++ show (is_prime n)
